Added RARE/freeRTr reference in mkdocs containerlab.dev

frederic-loui opened the pull request:
Hi, I added RARE/freeRtr reference. As per our discussion, instead of trying to create a new `kind`, i used `linux` `kind` instead. Feel free to correct accet or not  accept as si and modify accordingly if you feel contribution is "shaky"... Thanks !

hellt commented on lab-examples/rtr/000/rtr000.clab.yml:
```diff
@@ -0,0 +1,1715 @@
+
+
+
+
+
+
+
+<!DOCTYPE html>
+<html lang="en" data-color-mode="auto" data-light-theme="light" data-dark-theme="dark" data-a11y-animated-images="system">
+  <head>
+    <meta charset="utf-8">
+  <link rel="dns-prefetch" href="https://github.githubassets.com">
+  <link rel="dns-prefetch" href="https://avatars.githubusercontent.com">
```
@frederic-loui this looks like some accidentally committed file

frederic-loui replied:
commit `00a5e3a` should do the trick

hellt commented on docs/index.md:
```diff
@@ -58,36 +59,36 @@ This short clip briefly demonstrates containerlab features and explains its purp
     Although being kick-started by Nokia engineers, containerlab doesn't take sides and supports NOSes from other vendors and opensource projects.
-* **Lab orchestration**  
+* **Lab orchestration**
     Starting the containers and interconnecting them alone is already good, but containerlab packages even more features like managing lab lifecycle: [deploy](cmd/deploy.md), [destroy](cmd/destroy.md), [save](cmd/save.md), [inspect](cmd/inspect.md), [graph](cmd/graph.md) operations.
-* **Scaled labs generator**  
+* **Scaled labs generator**
     With [`generate`](cmd/generate.md) capabilities of containerlab it possible to define/launch CLOS-based topologies of arbitrary scale. Just say how many tiers you need and how big each tier is, the rest will be done in a split second.
-* **Simplicity and convenience**  
+* **Simplicity and convenience**
     Starting from frictionless [installation](install.md) and [upgrade](install.md#upgrade) capabilities and ranging to the behind-the-scenes [link wiring machinery](manual/network.md), containerlab does its best for you to enjoy the tool.
-* **Fast**  
+* **Fast**
     Blazing fast way to create container based labs on any Linux system with Docker.
-* **Automated TLS certificates provisioning**  
+* **Automated TLS certificates provisioning**
     The nodes which require TLS certs will get them automatically on boot.
-* **Documentation is a first-class citizen**  
+* **Documentation is a first-class citizen**
     We do not let our users guess by making a complete, concise and clean [documentation](https://containerlab.dev).
-* **Lab catalog**  
+* **Lab catalog**
    The "most-wanted" lab topologies are [documented and included](lab-examples/lab-examples.md) with containerlab installation. Based on this cherry-picked selection you can start crafting the labs answering your needs.
 
 ## Use cases
 
-* **Labs and demos**  
-    Containerlab was meant to be a tool for provisioning networking labs built with containers. It is free, open and ubiquitous. No software apart from Docker is required!  
-    As with any lab environment it allows the users to validate features, topologies, perform interop testing, datapath testing, etc.  
+* **Labs and demos**
+    Containerlab was meant to be a tool for provisioning networking labs built with containers. It is free, open and ubiquitous. No software apart from Docker is required!
+    As with any lab environment it allows the users to validate features, topologies, perform interop testing, datapath testing, etc.
     It is also a perfect companion for your next demo. Deploy the lab fast, with all its configuration stored as a code -> destroy when done. Easily and [securely share lab access](manual/published-ports.md) if needed.
-* **Testing and CI**  
```
your formatter removed trailing spaces here, but they are needed

frederic-loui replied:
Hi, I did not realised that my VScode formatter removed them as it was rendered correctly by `mkdocs`. I know now ! Thanks for the corrections !

hellt commented:
Hi @frederic-loui 
I think you have to add more to the lab example docs so that people would be able to deploy the lab.

I suggest you do the following:

1. create the necessary folders with `run` directory for the nodes you have in your lab, you can add empty files `.gitkeep` so that git checks in the dirs. I believe this is needed, since /run dir must exist before launching the nodes?
2. Provide steps to launch the lab. E.g. run `containerlab deploy <path to your topo file>` then connect with `telnet ...`

frederic-loui commented:
> 

1. I added `run` directory per `node` name in `rtr000` lab example
2. I actually-used [`frr`](https://containerlab.dev/lab-examples/frr01/) lab and essentiall the lab description and instruction are [here](https://github.com/rare-freertr/freeRtr-containerlab/blob/main/README.md)

I can certainly elaborate on containerlab `rtr000` lab page. As I might write more than one lab example, can you please indicate me which template I could use so that for consistency's sake I'll re-use the same lab template doc for each lab ? If you don't have one I can create one and stick to it for further lab documentation.

hellt commented:
@frederic-loui yes, this is a good one - https://github.com/rare-freertr/freeRtr-containerlab/blob/main/README.md

If you can refer to it from the lab example doc page for RARE it would be enough

frederic-loui commented:
> 

1- [One note] `.gitignore` contains `clab-*` regexp so i had to `-f`.
2- Reference added to lab example page. 

Thanks a lot for such (high frequency) reactivity !

hellt commented:
> 1- [note] `.gitignore` contains `clab-*` regexp so i had to `-f`.

to be honest, I think this qualifies a proper `rare` kind for this image. With a proper kind you will be able to create the needed directories automatically, so that users don't need to bother about it.

I am okay if you want to fist have a shortcut with using `linux` kind, but if you want to go the full `rare` kind route - we can pause this PR until you add the necessary changes.
You can have a look at this PR where xrd kind was added - https://github.com/srl-labs/containerlab/pull/1144/files#diff-c788dc041064f12abd92a7d5e1f4d66df17cd0458465874214a17a83d044e2ea

frederic-loui commented:
> > 1- [note] `.gitignore` contains `clab-*` regexp so i had to `-f`.
> 
> to be honest, I think this qualifies a proper `rare` kind for this image. With a proper kind you will be able to create the needed directories automatically, so that users don't need to bother about it.
> 
> I am okay if you want to fist have a shortcut with using `linux` kind, but if you want to go the full `rare` kind route - we can pause this PR until you add the necessary changes. You can have a look at this PR where xrd kind was added - https://github.com/srl-labs/containerlab/pull/1144/files#diff-c788dc041064f12abd92a7d5e1f4d66df17cd0458465874214a17a83d044e2ea

I did not want to impose you the creation of a new `kind` for `rare`, but as you kindly proposed, i cannot at least try to integrate it ! As you suggested, let's pause this PR and try to make things right at the very first time. I'm pretty sure that will provide a more awesome user experience with containerlab and RARE/freertr. I'll study the `xrd` `kind` commit above and get back to you ASAP.

Actually I would also be interested to add `type` to `kind` because of the nature of RARE/freeRtr having multiple dataplanes. (P4 BMv2, P4 TOFINO1/2, DPDK, XDP) do you have a commit with `type` implementation ? Each type would have it own daily built container from RARE/freeRTr `main` branch.

hellt commented:
@frederic-loui you can check `srl.go` for a kind that implements `type` property.
It is a free-formed string that inside the kind code you can parse and do the needed.

frederic-loui commented:
> > > 1- [note] `.gitignore` contains `clab-*` regexp so i had to `-f`.
> > 
> > 
> > to be honest, I think this qualifies a proper `rare` kind for this image. With a proper kind you will be able to create the needed directories automatically, so that users don't need to bother about it.
> > I am okay if you want to fist have a shortcut with using `linux` kind, but if you want to go the full `rare` kind route - we can pause this PR until you add the necessary changes. You can have a look at this PR where xrd kind was added - https://github.com/srl-labs/containerlab/pull/1144/files#diff-c788dc041064f12abd92a7d5e1f4d66df17cd0458465874214a17a83d044e2ea
> 
> I did not want to impose you the creation of a new `kind` for `rare`, but as you kindly proposed, i cannot at least try to integrate it ! As you suggested, let's pause this PR and try to make things right at the very first time. I'm pretty sure that will provide a more awesome user experience with containerlab and RARE/freertr. I'll study the `xrd` `kind` commit above and get back to you ASAP.
> 
> Actually I would also be interested to add `type` to `kind` because of the nature of RARE/freeRtr having multiple dataplanes. (P4 BMv2, P4 TOFINO1/2, DPDK, XDP) do you have a commit with `type` implementation ? Each type would have it own daily built container from RARE/freeRTr `main` branch.

Hi, I added `rare` `kind` following `xrd` `PR#1144`. 

One thing is that `all.go` has been removed. But i found `./clab/register.go` instead and the modified the file as follow:

```
./clab/register.go:	rare.Register(c.Reg)
```
I tested it and it worked as expected.
Essentially `rare` `kind` is creating `<clab-dir>/<node-short-name>/run` and bind it during node initialisation.
I issued also a `make test` and all test PASS'ed but to be honest I'm not sure if `rare` `kind` addition has been tested.

hellt commented on nodes/rare/rare.go:
```diff
@@ -0,0 +1,115 @@
+		return err
+	}
+
+	// when ignite runtime is in use
+	if n.vmChans != nil {
+		return <-n.vmChans.SpawnFinished
+	}
```
hi @frederic-loui 
do you need ignite runtime support? Or is this a copy leftover?

frederic-loui commented:
Actually I simply reused linux.go in order to better guarantee backwards
compatibility. And simply added "run"  folder creation and binding.

But i certainly don't need ignite runtime support




Le lun. 10 avr. 2023 à 21:09, Roman Dodin ***@***.***> a
écrit :

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In nodes/rare/rare.go
> <https://github.com/srl-labs/containerlab/pull/1315#discussion_r1161999605>
> :
>
> > +	// when ignite runtime is in use
> +	if n.vmChans != nil {
> +		return <-n.vmChans.SpawnFinished
> +	}
>
> hi @frederic-loui <https://github.com/frederic-loui>
> do you need ignite runtime support? Or is this a copy leftover?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/srl-labs/containerlab/pull/1315#pullrequestreview-1377986907>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJROB63NYFCRQXKLVKI6RDXARLHTANCNFSM6AAAAAAWVA6PGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hellt commented:
Thank you @frederic-loui 
great work!

Could you please check the edits I made, mainly I removed the ignite support and changed the lab example to point to your repo so that you don't have to maintain too repos yourself.
The idea is that if you have a working lab example maintained by RARE community, then no need to include it to containerlab repo, as you will be able to iterate faster that way.

Since I made changed to rare kind by removing the ignige support, I'd appreciate you testing it once again before I merge this.

Thanks again

steiler commented on nodes/rare/rare.go:
```diff
@@ -0,0 +1,66 @@
+	return createRAREFiles(n)
+}
+
+func createRAREFiles(node nodes.Node) error {
```
can we make this a method, please.
node.Config() is then availabe already as n.Cfg.

steiler replied:
While double-checking I see SROS is implemented in the same way. Should maybe also be altered seperately.

hellt replied:
agree, I did it in 1259b67

frederic-loui replied:
I just implemented `xrd``PR#1144` example provided by @hellt and got inspired with `crpd`(as It happened that i already had a `crpd` node). And yes, you are absolutely right this can be altered separately in a global approach. Because it is also present in :
* [cRPD](https://github.com/srl-labs/containerlab/blob/main/nodes/crpd/crpd.go#L76)

```
❯ grep -r create ./nodes | egrep "nodes.Node"
./nodes/rare/rare.go:func createRAREFiles(node nodes.Node) error {
./nodes/crpd/crpd.go:func createCRPDFiles(node nodes.Node) error {
./nodes/vr_sros/vr-sros.go:func createVrSROSFiles(node nodes.Node) error {
```

That would be awesome it you could correct this separately. (thanks !)

frederic-loui commented:
> Thank you @frederic-loui great work!
> 
> Could you please check the edits I made, mainly I removed the ignite support and changed the lab example to point to your repo so that you don't have to maintain too repos yourself. The idea is that if you have a working lab example maintained by RARE community, then no need to include it to containerlab repo, as you will be able to iterate faster that way.
> 
> Since I made changed to rare kind by removing the ignige support, I'd appreciate you testing it once again before I merge this.
> 
> Thanks again

First of all, _I_SHOULD_ thank you for this. All the work i've done is derived from `xrd` `PR#1144`. I've also rebuilt `containerlab` with `go` and checked that all is working properly. I presume you are got to go and close the PR. I'll now focus on creating various new labs. (srl+rare/freertr, crpd+rare/freertr etc. )
